include/actionlib/client/simple_action_client.h

```diff
@@ -535,15 +535,15 @@ void SimpleActionClient<ActionSpec>::handleTransition(GoalHandleT gh)
       switch (cur_simple_state_.state_) {
         case SimpleGoalState::PENDING:
         case SimpleGoalState::ACTIVE:
+          if (done_cb_) {
+            done_cb_(getState(), gh.getResult());
+          }
+
           {
             boost::mutex::scoped_lock lock(done_mutex_);
             setSimpleState(SimpleGoalState::DONE);
           }
 
-          if (done_cb_) {
-            done_cb_(getState(), gh.getResult());
-          }
-
           done_condition_.notify_all();
```

Comment:
To be honest, I don't think there is an issue in actionlib:
Actually, done-callbacks might expect that the current state is indeed `DONE` when they are called (I would do so).
The done_condition_.notify_all() ensures that threads waiting on this condition are only woke up when both the state changed and the callback was executed. So, from my point of view, everything is fine here.
I think we have another issue in MoveIt.

Reply:
@rhaschke Please see the docker container in #155 to reproduce the actionlib issue from the actionlib_tutorials. To check `ros::ok()` the thread is woken up every 0.1 seconds so it doesn't wait for the notification:
https://github.com/ros/actionlib/blob/c89e440519eeb88a5005065a32330655c9450b8e/include/actionlib/client/simple_action_client.h#L602-L603

I agree, that the client could expect the goal state to be `DONE`, so the fix isn't that simple...

Reply:
Thanks @jschleicher for this clarification! This was the point I was missing.
Probably, there should be an extra flag in actionlib that eventually indicates the `DONE` state - after the `done_cb_` was executed.
Great error analysis by the way!

Reply:
I've added a test to check `getState()` inside the callback. The state returned there is based on comm_state_:
https://github.com/ros/actionlib/blob/c89e440519eeb88a5005065a32330655c9450b8e/include/actionlib/client/simple_action_client.h#L346
So I still propose the original fix.

Reply:
> To be honest, I don't think there is an issue in actionlib:
> Actually, done-callbacks might expect that the current state is indeed `DONE` when they are called (I would do so).
> The done_condition_.notify_all() ensures that threads waiting on this condition are only woke up when both the state changed and the callback was executed. So, from my point of view, everything is fine here.
> I think we have another issue in MoveIt.

Couldn't there be spurious wakeups either way?

Reply:
> Couldn't there be spurious wakeups either way?

Yes, but the check of `cur_simple_state_` is still false in case of a spurious wake-up, so the loop continues: https://github.com/ros/actionlib/blob/c89e440519eeb88a5005065a32330655c9450b8e/include/actionlib/client/simple_action_client.h#L592

Reply:
Note that this PR is not about race-conditions or unstable timing, but the attached test-case *reliably fails* without the fix.